src/detect-engine.c

```diff
@@ -2164,6 +2179,45 @@ uint8_t DetectEngineInspectBufferGeneric(DetectEngineCtx *de_ctx, DetectEngineTh
     }
 }
 
+// wrapper for both DetectAppLayerInspectEngineRegister and DetectAppLayerMpmRegister
+// with cast of callback function
+void DetectAppLayerMultiRegister(const char *name, AppProto alproto, uint32_t dir, int progress,
+        InspectionMultiBufferGetDataPtr GetData, int priority, int tx_min_progress)
+{
+    DetectAppLayerInspectEngineRegisterAux(
+            name, alproto, dir, progress, DetectEngineInspectMultiBufferGeneric, NULL, GetData);
+    DetectAppLayerMpmMultiRegister(name, dir, priority, PrefilterMultiGenericMpmRegister, GetData,
+            alproto, tx_min_progress);
+}
+
+uint8_t DetectEngineInspectMultiBufferGeneric(DetectEngineCtx *de_ctx,
+        DetectEngineThreadCtx *det_ctx, const DetectEngineAppInspectionEngine *engine,
+        const Signature *s, Flow *f, uint8_t flags, void *alstate, void *txv, uint64_t tx_id)
+{
+    uint32_t local_id = 0;
+    const DetectEngineTransforms *transforms = NULL;
+    if (!engine->mpm) {
+        transforms = engine->v2.transforms;
+    }
+
+    while (1) {
+        InspectionBuffer *buffer = engine->v2.GetMultiData(
+                det_ctx, transforms, f, flags, txv, engine->sm_list, local_id);
+
+        if (buffer == NULL || buffer->inspect == NULL)
+            break;
+
+        const bool match = DetectEngineContentInspection(de_ctx, det_ctx, s, engine->smd, NULL, f,
+                buffer->inspect, buffer->inspect_len, buffer->inspect_offset,
+                DETECT_CI_FLAGS_SINGLE, DETECT_ENGINE_CONTENT_INSPECTION_MODE_STATE);
```

Comment:
this `DETECT_CI_FLAGS_SINGLE` limits this to non-streaming buffers. This may be intentional, but it's something to keep in mind.

Reply:
additionally, it's hardcoding setting `DETECT_ENGINE_CONTENT_INSPECTION_MODE_STATE`. This can be fine, but it means that this logic is limited to state keywords. I suppose it means the function naming should reflect it.

Reply:
> this DETECT_CI_FLAGS_SINGLE limits this to non-streaming buffers. This may be intentional, but it's something to keep in mind.

That looks intentional IIRC

> additionally, it's hardcoding setting DETECT_ENGINE_CONTENT_INSPECTION_MODE_STATE. This can be fine, but it means that this logic is limited to state keywords. I suppose it means the function naming should reflect it.

🤔 What are "state keywords" ?

Reply:
`git show | grep -C2 DetectEngineContentInspection` shows always usage of `DETECT_ENGINE_CONTENT_INSPECTION_MODE_STATE`. in the deleted lines

Reply:
This helped me find that `DetectEngineContentInspectionBuffer` now got unused

Reply:
Probably better to use it, as it "solves" the DETECT_CI_FLAG "issue" by using the flag from InspectionBuffer